### PR TITLE
Add missing conversion for to_sparse.sparse_dim

### PR DESCRIPTION
**Pull request by alexsamardzic**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92022
* __->__ #92006

cc @nikitaved @pearu @cpuhrsch @amjames @bhosmer

### CODE REVIEW

**Comment by alexsamardzic** [comment]
Just a small fix for following snippet, that causes `NotImplementedError`:
```python
coo = torch.sparse_coo_tensor(((0, 1), (0, 1)), (1, 1))
coo2 = coo.to_sparse(sparse_dim=2)
```

File: aten/src/ATen/native/TensorConversions.cpp [hunk 1]
```diff
@@ -1752,14 +1752,19 @@ Tensor sparse_compressed_to_sparse_csc(const Tensor& self) {
...
+
 Tensor sparse_compressed_to_sparse(const Tensor& self, int64_t sparse_dim) {
-  TORCH_CHECK(sparse_dim > 0, "sparse_dim must be >0");
-  TORCH_CHECK(sparse_dim <= 2,
-              "sparse_dim must be less than or equal to 2");
-  // TODO: implement coo.to_sparse(sparse_dim) and then use
-  // return self.to_sparse().to_sparse(sparse_dim);
   TORCH_CHECK(
-      sparse_dim == 2, "sparse dim 1 is not supported by sparse_compressed_to_dense");
+      sparse_dim == 2, "sparse dim argument must be 2 for sparse_compressed_to_sparse");
```
Comment (nikitaved) [inline aten/src/ATen/native/TensorConversions.cpp:1767]: if `self.sparse_dim() == sparse_dim`, can't we just return `self.clone()`?

File: aten/src/ATen/native/TensorConversions.cpp [hunk 2]
```diff
@@ -1701,14 +1701,19 @@ Tensor sparse_compressed_to_sparse_csc(const Tensor& self, c10::optional<int64_t
   return self;
 }
 
+Tensor sparse_coo_to_sparse(const Tensor& self, const int64_t sparse_dim) {
+  TORCH_CHECK(
+      sparse_dim >= 0 && sparse_dim <= self.sparse_dim(), "sparse_dim argument must be between 0 and number of sparse dimensions of original tensor");
+  if (sparse_dim != self.sparse_dim()) {
+    TORCH_WARN(
+        "sparse_dim argument ", sparse_dim, " ignored in sparse_coo_to_sparse, number of sparse dimensions kept from original tensor instead");
```
[the remark this answers is not in the document]
Comment (alexsamardzic) [reply aten/src/ATen/native/TensorConversions.cpp]: Changed.

**Review by pearu — APPROVED** [review]
LGTM! Thanks, @alexsamardzic!

Btw, the coo->coo conversion with sparse_dim change can be defined as follows. Consider `coo` tensor with `0 < coo.sparse_dim() <= coo.ndim`.

1. if `coo.sparse_dim() == sparse_dim`, return `coo` (this PR).
2. if `coo.sparse_dim() < sparse_dim`, then compute `coo1 = coo.values().to_sparse(sparse_dim - coo.sparse_dim())` and return `torch.sparse_coo_tensor("outer product of coo and coo1 indices", coo1.values())` or similar.
3. if `coo.sparse_dim() > sparse_dim`, then return `torch.sparse_coo_tensor(coo.indices()[:sparse_dim], "sparse_coo_tensor(coo.indices()[sparse_dim:], coo.values()).to_dense()")` or similar.

where expressions in double quotes and "or similar" mean that the correct details must be figured out... But the main point is that supporting different sparse_dim might not be so hard.

**Comment by nikitaved** [comment]
To add on what @pearu mentioned, this is similar to what we do in `mul(sparse, dense)` which does broadcast.

**Comment by alexsamardzic** [comment]
Thanks for the suggestions.  The PRs in this stack are just quick fixes to issues encountered while looking at the status of sparse-to-sparse conversions.   In general, I agree that it would be ideal to be able at some point to change numbers of batch/sparse/dense dims (whatever combinations applicable to given input/output layouyts) at will.

**Review by cpuhrsch — APPROVED** [review]

File: aten/src/ATen/native/TensorConversions.cpp [hunk 3]
```diff
@@ -1695,14 +1695,17 @@ Tensor sparse_compressed_to_sparse_csc(const Tensor& self, c10::optional<int64_t
   return self;
 }
 
+Tensor sparse_coo_to_sparse(const Tensor& self, const int64_t sparse_dim) {
+  TORCH_CHECK(
+      sparse_dim >= 0 && sparse_dim <= self.sparse_dim(), "sparse_dim argument for sparse_coo_to_sparse must be between 0 and number of sparse dimensions of original tensor");
```
Comment (nikitaved) [inline aten/src/ATen/native/TensorConversions.cpp:1700]: Isn't this check a bit redundant? The user will be really angry to hit the first, and then, after potentially a couple of ours, the second :)

Comment (alexsamardzic) [reply aten/src/ATen/native/TensorConversions.cpp]: Removed (this PR initially proposed the second check as warning instead).

**Review by pearu — APPROVED** [review]
LGTM!

**Pull request closed without merging**